### ISSUE TITLE
hostnames in /etc/hosts containing an underscore are not valid

### ISSUE THREAD

**Issue by feelotraveller** [systemd/systemd#11500]
**systemd version the issue has been seen with**
240.275  

**Used distribution**
Manjaro

**Expected behaviour you didn't see**
Hostnames containing underscores in /etc/hosts to be regarded as valid.

**Unexpected behaviour you saw**
systemd-resolved[438]: /etc/hosts:90: hostname "222-33544_999.pub.adfirmative.com" is not valid, ignoring.
systemd-resolved[438]: /etc/hosts:6068: hostname "ghm_bulgaria.hit.gemius.pl" is not valid, ignoring.
systemd-resolved[438]: /etc/hosts:13151: hostname "philadelphia_cbslocal.us.intellitxt.com" is not valid, ignoring.
systemd-resolved[438]: /etc/hosts:13386: hostname "___id___.c.mystat-in.net" is not valid, ignoring.

Edit: not sure how to bypass the markdown styling on the fourth address, there are triple underscores either side of 'id'.

**Steps to reproduce the problem**
Have a hostname in /etc/hosts that includes an underscore and boot system.  Find report with journalctl -b0 -p4.

I am using the list of around 14000 hostnames to block (0.0.0.0) from [MVPS](http://winhelp2002.mvps.org/hosts.htm).  These are the only 4 hosts that contain underscores as part of their names.

I have pinged the addresses without problem and resolvectl query has succeeded in providing a proper response so the problem does not appear to be with the addresses themselves.

The problem has not occurred previously with the same hosts file.  Most recent previous systemd version installed was 239.370.

Happy to provide further information if needed.  Beware my newness!

**Comment by zdzichu** [comment]
Hostname with underscores and invalid according to RFC, see http://domainkeys.sourceforge.net/underscore.html

**Comment by poettering** [comment]
hmm, not sure what to do with this one... better to be RFC compliant or compatible? Usually I'd say compat wins, but this sounds nichey enough to side with RFC compliance?

**Comment by kragniz** [comment]
I'd personally say to side with compatibility, unless allowing this causes problems somewhere else

**Comment by b1tninja** [comment]
#11765 

Ignoring the RFC is obviously a slippery slope but maybe some leniency for valid Netbios/computer names seems acceptable. Underscores also are used as a prefix for SRV records... Technically the RFC says "should", not a strict requirement. And by their definition things like 7.3.3.1.in-addr.arpa would be invalid-- no idea why you'd find it in a hosts file, but just food for thought.

@feelotraveller any idea if the IDN (xn--) would resolve properly? Maybe nonstandard names could be translated?

https://www.ietf.org/rfc/rfc822.txt
https://www.ietf.org/rfc/rfc1035.txt (page 7)
https://support.microsoft.com/en-us/help/909264/naming-conventions-in-active-directory-for-computers-domains-sites-and

**Comment by feelotraveller** [comment]
IDN (xn--) does nothing here since the underscore is a valid ASCII character (code 242).

I did not find a way to translate the addresses to strict RFC - wildcards not being valid in the hosts file -  but have 'resolved' ;) to bear with the niggle of the warnings.  Someone who knows more may know how to achieve this though.

Although the warnings persist the addresses are correctly referenced and blocked.  For example
```
getent hosts philadelphia_cbslocal.us.intellitxt.com
0.0.0.0         philadelphia_cbslocal.us.intellitxt.com
```

An alternative approach would be to block elsewhere, say at the router or using iptables, but I find it convenient to do it this way.

**Comment by MatMercer** [comment]
Posting this here in case anyone faces the same issue. We use `example_hostname` at `/etc/hosts`.

At Windows/Mac it works file, but with my Arch Linux machine it required a workaround.

```sh
sudo systemctl disable --now systemd-resolved
sudo pacman -S dnsmasq # after this, hostnames with _ started working immediately 
```

This costed me 2 hours of debugging because I was blocked from local development. I feel like this is another strong evidence it's best to go with the "compatibility" route, because I only found the issue after I tested `examplehostname` and I was like "ooohhh so it's that".
